### PR TITLE
Configparams

### PR DESCRIPTION
**Pull request by dzmarshall**
Make sure that the configuration parameters are getting set as the correct type and being written when prompted at deploy.

### CODE REVIEW

**Comment by RobertFischer** [comment]
Much nicer pull request; I appreciate it!

File: app/models/deployment.rb [hunk 1]
```diff
@@ -36,7 +36,7 @@ def check_stage_ready
       errors.add('stage', 'is not ready to deploy') unless self.stage.deployment_possible?
 
       self.stage.prompt_configurations.each do |conf|
-        errors.add('base', "Please fill out the parameter '#{conf.name}'") unless !prompt_config.blank? && !prompt_config[conf.name.to_sym].blank?
+        errors.add('base', "Please fill out the parameter '#{conf.name}'") unless !prompt_config.blank? && !prompt_config[conf.name].blank?
```
Comment (RobertFischer) [inline app/models/deployment.rb:39]: Was there a problem with the `.to_sym` call?

File: lib/capsize/deployer.rb [hunk 2]
```diff
@@ -123,6 +123,28 @@ def self.cvs_root_defintion?(val)
...
+      val = parameter.value
+
+      val = deployment.prompt_config[name] if parameter.prompt?
+
+      return nil if val.nil?
+      val.strip!
+      return set_string_parameter(name, val) if val =~ /\A\d+\D+/
+
+      case val[0]
+      when ":", "%", "[", "{", /\d/
```
Comment (RobertFischer) [inline lib/capsize/deployer.rb:142]: `\d`?  So the value `foo2` isn't a string parameter?

File: lib/capsize/deployer.rb [hunk 3]
```diff
@@ -123,6 +123,28 @@ def self.cvs_root_defintion?(val)
...
+        "set :#{name}, #{val}"
+      else
+        set_string_parameter(name, val)
+      end
+    end
+
+    def set_string_parameter(name, val)
+      "set :#{name}, '#{val}'"
```
Comment (RobertFischer) [inline lib/capsize/deployer.rb:150]: Why does this deserve its own method, when the other branch (line 138) is done inline?

File: lib/capsize/deployer.rb [hunk 4]
```diff
@@ -157,7 +179,7 @@ def write_stage
           f.puts "role :#{role.name}, %w{#{find_host_user(@project)}@#{role.host.name}}"
         end
         @stage.configuration_parameters.each do |parameter|
-          f.puts "set :#{parameter.name}, \"#{parameter.value}\""
+          f.puts print_parameter(parameter)
```
Comment (RobertFischer) [inline lib/capsize/deployer.rb:191]: These changes are nice — always better to have meaningful, self-documenting method calls than inline constants. Good job.

File: test/unit/capsize_deployer_test.rb [hunk 5]
```diff
@@ -193,17 +193,17 @@ def test_handling_of_prompt_configuration
     assert stage_with_prompt.deployment_possible?, stage_with_prompt.deployment_problems.inspect
 
     # add a config value that wants a promp
-    stage_with_prompt.configuration_parameters.build(:name => 'password', :prompt_on_deploy => 1).save!
+    stage_with_prompt.configuration_parameters.create(:name => 'password', :prompt_on_deploy => 1)
```
Comment (RobertFischer) [inline test/unit/capsize_deployer_test.rb]: Shouldn't this be `create!` to stick with the existing `save!` semantics? You're also not checking the return value, which means you are silently discarding an error condition.

File: test/unit/capsize_deployer_test.rb [hunk 6]
```diff
@@ -193,17 +193,17 @@ def test_handling_of_prompt_configuration
...
-    deployment = create_new_deployment(:stage => stage_with_prompt, :task => 'deploy', :prompt_config => {:password => '123'})
+    deployment = create_new_deployment(:stage => stage_with_prompt, :task => 'deploy', :prompt_config => {'password' => 'pass1234'})
 
     write_deployer_files(deployment)
-    file = File.open(rooted("#{stage_with_prompt.project.capsize_project_name}/#{stage_with_prompt.name}.rb"))
+    file = File.open(rooted("#{@project.capsize_project_name}/#{stage_with_prompt.name}.rb"))
     contents = file.read
 
-    assert_match(/set :password, "123"/, contents)
+    assert_match(/set :password, 'pass1234'/, contents)
```
Comment (RobertFischer) [inline test/unit/capsize_deployer_test.rb:206]: This looks extremely suspicious. It's worth a comment in the code to describe what's going on — you have a hardcoded password?

File: test/unit/capsize_deployer_test.rb [hunk 7]
```diff
@@ -261,6 +261,24 @@ def test_stage_file_is_written
...
+    @stage.project.configuration_parameters.create(:name => 'keep_releases', :value => '8')
+    @stage.project.configuration_parameters.create(:name => 'linked_dirs', :value => '%w{ made_up_dir }')
+
+    write_deployer_files(deployment)
+    file = File.open(rooted("#{@stage.project.capsize_project_name}/deploy.rb"))
+    contents = file.read
+
+    assert_match /set :app_name, 'madeupname'/, contents
+    assert_match /set :keep_releases, 8/, contents
+    assert_match /set :linked_dirs, %w{ made_up_dir }/, contents
```
Comment (RobertFischer) [inline test/unit/capsize_deployer_test.rb]: You probably want to ensure these are on their own line (using `^` and `$` and the `m` regex modifier, possibly with some `\s*` tossed in for good measure). Also, add some trickier cases, such as `"23skidoo"`, `"foo2"`, `nil`, etc. For bonus points, make a single helper method in this test class which takes a name, a value, and an expected string, and then a series of tests which call into that helper method.

**Comment by RobertFischer** [comment]
Some weird stuff, some suspicious stuff; rejecting the ticket. Please respond to the comments, rework the branch, and then reopen this pull request.

File: app/models/deployment.rb (hunk 1 above)
Comment (dzmarshall) [inline app/models/deployment.rb:39]: The old way of doing things had the configuration parameter names converted into symbols and saved into the configuration hash, now we read them from the db as strings.  This call was using `.to_sym` to look for a symbol version of a parameter name saved as a string.

File: lib/capsize/deployer.rb (hunk 2 above)
Comment (dzmarshall) [inline lib/capsize/deployer.rb:142]: `\d` and the rest of the when statement is checking the first character of `val`, so `foo2` would be a string parameter as well as `2foo` which would be caught by the above guard clause.

File: lib/capsize/deployer.rb (hunk 3 above)
Comment (dzmarshall) [inline lib/capsize/deployer.rb:150]: I used it twice, once under `else` and once in the guard clause.  Though I have thought of something I might have missed in this method so I will probably be making a method for 138 as well.

File: test/unit/capsize_deployer_test.rb (hunk 5 above)
Comment (dzmarshall) [inline test/unit/capsize_deployer_test.rb]: Yep, I'll change that.

File: test/unit/capsize_deployer_test.rb (hunk 6 above)
Comment (dzmarshall) [inline test/unit/capsize_deployer_test.rb:206]: This is just a test, all its doing is saying that the password parameter value should be prompted on deploy, then on line 200 I set the password parameter in the `prompt_config` hash to some arbitrary string to simulate the user entering in the value at deploy and then line 206 I check to make sure that `:password` and the arbitrary string are written in the deploy file.

File: test/unit/capsize_deployer_test.rb (hunk 7 above)
Comment (dzmarshall) [inline test/unit/capsize_deployer_test.rb]: OK, I'll add more to this.

File: app/models/deployment.rb (hunk 1 above)
Comment (RobertFischer) [inline app/models/deployment.rb:39]: Perfect.

File: lib/capsize/deployer.rb (hunk 2 above)
Comment (RobertFischer) [inline lib/capsize/deployer.rb:142]: Good point — missed that it was `val[0]` instead of `val`.

File: lib/capsize/deployer.rb (hunk 3 above)
Comment (RobertFischer) [inline lib/capsize/deployer.rb:150]: Yeah, symmetry would be nice, just so it's clear that you're not doing something fundamentally different.

File: test/unit/capsize_deployer_test.rb (hunk 6 above)
Comment (RobertFischer) [inline test/unit/capsize_deployer_test.rb:206]: Ah, okay. That's fine, then. Doesn't need a comment.

**Pull request merged**